WalletWasabi/Tor/Http/IHttpClient.cs

```diff
@@ -15,5 +17,51 @@ public interface IHttpClient
 
 		/// <inheritdoc cref="HttpClient.SendAsync(HttpRequestMessage, CancellationToken)"/>
 		Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken token = default);
+
+		Task<HttpResponseMessage> SendAsync(HttpMethod method, Uri requestUri, HttpContent? content = null, CancellationToken token = default)
+		{
+			var httpRequestMessage = new HttpRequestMessage(method, requestUri);
+
+			if (content is { })
+			{
+				httpRequestMessage.Content = content;
+			}
+
+			return SendAsync(httpRequestMessage);
+		}
+
+		/// <summary>
+		/// Sends HTTP request (up to <paramref name="retry"/> times) to get an HTTP response.
+		/// </summary>
+		/// <param name="method">HTTP method to use to send the HTTP request.</param>
+		/// <param name="requestUri">URI to send HTTP request to. Cannot be relative URI.</param>
+		/// <param name="retry">Maximum number of attempts to do to get HTTP response.</param>
+		/// <exception cref="OperationCanceledException">When <paramref name="token"/> is used to cancel operation.</exception>
+		public async Task<HttpResponseMessage?> SendAndRetryAsync(HttpMethod method, Uri requestUri, int retry = 2, HttpContent? content = null, CancellationToken token = default)
+		{
+			HttpResponseMessage? response = null;
+
+			while (retry-- > 0)
+			{
+				response?.Dispose();
+				token.ThrowIfCancellationRequested();
+				response = await SendAsync(method, requestUri, content, token).ConfigureAwait(false);
+
+				if (response.StatusCode == HttpStatusCode.OK)
```

Comment:
Does this assumes that the only valid HTTP response is `HttpStatusCode.OK`? If so, then that's not good. HTTP status codes are handled in Wasabi in various ways. For example if a round aborted or alice kicked from the coinjoin, then the server gives back `NotFound`. So this would force Tor to make another request hoping that Alice would somehow brought back to life or something? I don't really understand this. If we receive a HTTP status code, then that's proof there's already proper communication built out between the server and the client, otherwise we wouldn't be able to receive data (the status code) from the server.

Reply:
This is to summarize how we got here first:

* `TorHttpClientExtensions` used to exist [[link](https://github.com/zkSNACKs/WalletWasabi/commit/568515b7e7761741a9f8ac987de48d17ce158f97#)]
* `TorHttpClientExtensions.SendAndRetryAsync` used to have `HttpStatusCode expectedCode` parameter that was always set to `HttpStatusCode.OK` and for that reason it was removed [[link](https://github.com/zkSNACKs/WalletWasabi/commit/568515b7e7761741a9f8ac987de48d17ce158f97#diff-e18aa2c9cc817ea07aa99f56ee6157aa7d2139a5e687ad18aaa9fce93e3355e1L15)]
* `TorHttpClientExtensions.SendAndRetryAsync` was moved to `ITorHttpClient`

and finally

* This PR copies `ITorHttpClient.SendAndRetryAsync` to `IHttpClient` (for absolute URIs, not relative as in `ITorHttpClient.SendAndRetryAsync`). As I wrote above this is to make this PR reasonably big.

So this is history in a nutshell. 

-> I don't actually change the behavior here or do you see any change introduced by me?

Where is `SendRetryAsync` used?

* `IHttpClient.SendRetryAsync` is used only in BackendTests.
* `ITorHttpClient.SendRetryAsync` is used by `WasabiClient` class.

> Does this assumes that the only valid HTTP response is `HttpStatusCode.OK`?

Yes, it does. Still it can be easily changed, if it is actually needed by a caller.

> So this would force Tor to make another request hoping that Alice would somehow brought back to life or something?

I don't see any use of any `SendRetryAsync` in Coinjoin code. So no, I don't think so.

Does this answer your questions?

Reply:
Not really. My most relevant (meta) question is if you acknowledge this is a bug and intend fix it or I need to sit down, get involved and spend some considerable time on it?

Reply:
I understand what you are saying but I can't fix it properly as it is used on many places where I don't know the code well enough. Also, I don't know what specific issues `SendRetryAsync` is supposed to fix. And most certainly I don't know all HTTP status codes that may be returned for individual http calls to backend/coinjoin. That's intractable for me.

If we can split the task to smaller issues, I can probably fix it. But changing `SendRetryAsync` by me can break something.

Reply:
It can be improved, there are cases when it makes no sense to retry the request. 

How about just having one SendAndRetryAsync in IHttpClient with relative URI-s? 
We are more likely to use things like this:
```C#
string relativeUri = $"/api/v{ApiVersion}/btc/batch/synchronize?bestKnownBlockHash={bestKnownBlockHash}&maxNumberOfFilters={count}";
if (estimateMode is { })
{
	relativeUri = $"{relativeUri}&estimateSmartFeeMode={estimateMode}";
}
```

Reply:
> How about just having one SendAndRetryAsync in IHttpClient with relative URI-s?

`IHttpClient` is not ready for relative URIs now as ClearnetHttpClient is only wrapper over .NET's HttpClient. If we decide, this is the correct way, then this must be done first I guess.

The key question is whether `IHttpClient` should work with relative URIs in the first place. It looks like it's nicer API to work with as a developer but it has some substantials cons.

When you have TorHttpClient that accepts base URI, then:

* It means that switching to a fallback URI feels weird and you actually use TorHttpClient as if it has a method for absolute URI. 
* You need to create always a new TorHttpClient instance when you want to make an HTTP request to a new base URI.
* It complicates paralelization efforts as you need to pool with regard to base URI too.

> We are more likely to use things like this:

That may be true but a simple wrapper for IHttpClient can achieve that too and you would still have general ITorHttpClient.

> It can be improved, there are cases when it makes no sense to retry the request.

It would be great to identify those places and put them in an issue so that it can be addressed. 

Feel free to comment if I'm not missing something.

Reply:
According to our discussion: 

- We will go with the parallelization and the simple pool approach that requires absolute URI. So that is the way to go and your wrapper idea will be helpful.
- SendAndRetry was a hacky solution to some issues in the past, so it can be removed.